File: src/insert-delete-getrandom-o1-duplicates-allowed/insert-delete-getrandom-o1-duplicates-allowed.rs
// https://leetcode.com/problems/insert-delete-getrandom-o1-duplicates-allowed

use std::collections::HashMap;
use std::collections::HashSet;
use rand::rngs::ThreadRng;
use rand::seq::SliceRandom;

#[derive(Debug)]
struct RandomizedCollection {
    values: Vec<i32>,
    // val -> (idx, count)
    index: HashMap<i32, HashSet<usize>>,
    thread_rng: ThreadRng,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RandomizedCollection {

    fn new() -> Self {
        let values = vec![];
        let index = HashMap::new();
        let thread_rng = rand::thread_rng();

        Self { values, index, thread_rng }
    }
    
    fn insert(&mut self, val: i32) -> bool {
        match self.index.get_mut(&val) {
            Some(hset) => {
                assert!(hset.insert(self.values.len()));
                self.values.push(val);
                false
            },
            None =>  {
                self.index.insert(val, HashSet::from([self.values.len()]));
                self.values.push(val);
                true
            }
        }
    }

    fn val_idx_pop_next(&self, val: &i32) -> Option<usize> {
        if let Some(hset) = self.index.get(val) {
            let idx = hset.iter().next()
                .expect("Hset present with no indices")
                .to_owned();

            return Some(idx);
        }

        None
    }
    
    fn remove(&mut self, val: i32) -> bool {
        if let None = self.index.get(&val) {
            return false;
        }

        let idx = self.val_idx_pop_next(&val)
            .expect("Index had val entry, but did not get index to pop");
        let last = self.values.last()
            .expect("Fail to get last")
            .to_owned();
        let last_hset = self.index.get_mut(&last)
            .expect("Fail to get last set");
        assert!(last_hset.remove(&(self.values.len() - 1)));
        last_hset.insert(idx);
        assert!(self.values.swap_remove(idx) == val);

        let hset = self.index.get_mut(&val)
            .expect("Did not get hset for val");
        hset.remove(&idx);

        true
    }
    
    fn get_random(&mut self) -> i32 {
        self.values.as_slice()
            .choose(&mut self.thread_rng)
            .expect("Attempt to get_random() while empty")
            .to_owned()
    }
}

fn main() {
    let mut obj = RandomizedCollection::new();
    assert!(obj.insert(1));
    println!("obj = {:?}", obj);
    assert!(obj.get_random() == 1);
    assert!(obj.remove(1));
    assert!(!obj.remove(2));

    assert!(obj.insert(0));
    assert!(obj.insert(1));
    assert!(obj.insert(2));
    for _ in 0..100 {
        let val = obj.get_random();
        assert!(val == 0 || val == 1 || val == 2)
    }

    assert!(obj.remove(1));
    assert!(!obj.remove(1));
    for _ in 0..100 {
        let val = obj.get_random();
        assert!(val == 0 || val == 2)
    }
    assert!(obj.remove(0));
    assert!(obj.remove(2));
 
    println!("✅ All tests passed");
}

File: src/insert-delete-getrandom-o1-duplicates-allowed/insert-delete-getrandom-o1-duplicates-allowed.py
import random

class RandomizedCollection(object):

    def __init__(self):
        self.values = list()
        self.indexmap = dict()

    def insert(self, val):
        """
        :type val: int
        :rtype: bool
        """
        idx = len(self.values)
        exists = val in self.indexmap
        if not exists:
            self.indexmap[val] = set()
        self.indexmap[val].add(idx)
        self.values.append(val)
        return not exists


    def remove(self, val):
        """
        :type val: int
        :rtype: bool
        """
        if val not in self.indexmap:
            return False
        last_idx = len(self.values) - 1
        last = self.values[last_idx]
        last_set = self.indexmap[last]

        rm_set = self.indexmap[val]
        rm_idx = next(iter(rm_set))

        if last == val:
            rm_set.remove(last_idx)
        else:
            # Update index hashsets
            rm_set.remove(rm_idx)
            last_set.remove(last_idx)
            last_set.add(rm_idx)
            # swap-remove
            self.values[last_idx], self.values[rm_idx] = \
                    self.values[rm_idx], self.values[last_idx]

        self.values.pop()
        if len(rm_set) == 0:
            del self.indexmap[val]

        return True

    def getRandom(self):
        """
        :rtype: int
        """
        idx = random.randint(0, len(self.values) - 1)
        return self.values[idx]


if __name__ == "__main__":
    randset = RandomizedCollection()
    ops = [
        ("insert",      (1,),   True),
        ("insert",      (1,),   False),
        ("remove",      (2,),   False),
        ("insert",      (2,),   True),
        ("getRandom",   (),     "rand"),
        ("remove",      (1,),   True),
        ("remove",      (1,),   True),
        ("remove",      (1,),   False),
        ("getRandom",   (),     2),
    ]
    for (method, args, out) in ops:
        method = getattr(randset, method)
        ret = method(*args)
        if out != "rand":
            assert out == ret, \
                f"out {out} != ret {ret}"
    print("✅ All tests passed")

